keps/sig-apps/sidecarcontainers.md

```diff
@@ -67,7 +67,7 @@ Solve issues so that they don't require application modification:
 
 ## Non-Goals
 
-Allowing multiple containers to run at once during the init phase. //TODO See if we can solve this problem with this proposal
+Allowing multiple containers to run at once during the init phase - this could be solved using the same principal but can be implemented separately. //TODO should this be a non-goal or an extension? is there enough support for this?
```

Comment:
Maybe not quite what you were asking: Do we have any use-cases or even hypotheticals that speak to the intersection of this design and init containers?  E.g. Do init containers need service-mesh functionality?   I would assume they do.

In other words, should sidecars start before init containers?  And then, do we have sequencing within init containers?  Concretely, consider something like Istio which has an init container with privilegs to write iptables rules, then a sidecar which processes net traffic.  It seems to me that the sidecar would be needed for any user-provided init containers.  I don't know if this has come up in real life yet.

How do we think that holds up in general?

@louiscryan @geeknoid @costinm for some perspective

Reply:
@thockin So this was brought up by istio specifically, see https://github.com/kubernetes/enhancements/issues/753 and https://github.com/istio/cni/issues/77. I think init containers needing service mesh access seems like a fairly valid use case. I think more generally the idea of an init container needing a sidecar isn't unheard of, e.g I have an init container that performs a db schema migration but it should go through mysql proxy.

I think you can also assume that some sidecars may need the init containers to run before they start such as the example of istio with init doing iptables and sidecar then starting for traffic.

You could potentially achieve this using what we've already suggested but adapting it for init e.g 
```
initContainers:
- name: iptables
  command: ['set up some iptables]
- name: proxy
  command: ['proxy-stuff']
  sidecar: true
- name: myinitapp
  command: ['do some init stuff that requires the proxy']
```
The difference being it would follow order like init does already, but when it gets to the sidecar it could wait for it to be ready instead of complete, you then clean up the init sidecars at the end of the init phase.
This has the advantage that it keeps the UX relatively simple, the init phase is still logically separated, while hopefully enabling the use case.

Reply:
I agree with the need for the functionality, but something feels wonky UX-wise about boot order jumping between the containers and initContainers sections.

Would it make sense to have a more explicit DAG of boot operations?

Reply:
@vllry, what do you mean by boot order jumping? Perhaps I wasn't that clear but I was suggesting we still have all the init containers finishing before the containers start. You could just have init sidecars that run alongside the init containers for that phase only.

Reply:
What I meant was that having a sidecar toggle for initContainers seems clunky in the current config (the idea overall idea makes sense).

EG I might have a boot order like this:

```
initContainers:
- name: 1
  sidecar: true
  command: ...
- name: 3
  command: ...
containers:
- name: 2
  sidecar: true
- name: 4
```

It definitely works, but is awkward to read.

Reply:
I was suggesting that Init container sidecars would start and finish in the init phase, container sidecars would start after the init phase is over, so there would be no way to have a boot order like you described. I agree it would be confusing if this was the case.

Reply:
My vote would be to explore this just enough to be able to write a short section in the doc explaining how this could be extended to init containers in the future (at a similar level of implementation detail as the current section for implementing non-init sidecars).

We have clear, concrete use cases for non-init sidecars (Jobs, log/result shippers, sql proxy, Istio) and I don't think we should block addressing those. Non-init sidecar is already a pattern that's in use in the real world, and we just want to replace ugly workarounds (wrapper scripts, etc) with true support. Init sidecars are not in use because it's not possible even as a workaround.

Reply:
I think init containers should continue to be executed to completion strictly before sidecars are launched. The purpose of init containers is to perform actions that pertain to the Pod that need to happen prior to the main containers starting (e.g. setting up configuration files, preparing the file system, pulling data down). Launching run to completion containers subsequent to "main" containers might  be problematic. Perhaps in the future we could add init-sidecars?